### PR TITLE
Formatter expand/collapse/ignore empty XML tags

### PR DESCRIPTION
**Pull request by angelozerr**
Formatter expand/collapse/ignore empty XML tags

See https://github.com/redhat-developer/vscode-xml/issues/219

Signed-off-by: azerr <azerr@redhat.com>

### CODE REVIEW

**Comment by angelozerr** [comment]
This PR adds new format settings `"xml.format.formatEmptyElements"` which can takes string values:

 * none (do nothing like today)
 * collapse : collapse empty elements
 * expand : expand empty elements

Perhaps the setting should be renamed to `"xml.format.emptyElements"` ?

This PR requires the vscode xml PR https://github.com/redhat-developer/vscode-xml/pull/241 which provides this settings.

Here a demo:

FormatEmptyElements

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/settings/XMLFormattingOptions.java [hunk 1]
```diff
@@ -40,27 +40,39 @@
...
+	}
+
 	private static final String PRESERVE_EMPTY_CONTENT = "preserveEmptyContent";
 
+	/**
+	 * Options for formatting empty elements.
+	 *
+	 */
+	public static enum FormatEmptyElements {
+		expand, collapse, none;
```
Comment (fbricon) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/settings/XMLFormattingOptions.java:55]: needs javadoc to explain the differences

Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/settings/XMLFormattingOptions.java:96]: fixed

**Comment by angelozerr** [comment]
> Does the formatting respect XSD definitions like isNillable?

No, but I think we should manage in an another issue.

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/settings/XMLFormattingOptions.java [hunk 2]
```diff
@@ -40,27 +40,80 @@
...
+	 * will be formatted to :
+	 * 
+	 * <pre>
+	 * {@code
+	 * <example />
+	 * }
+	 * </pre>
+	 * 
+	 * </li>
+	 * <li>{@link #none} : keep the original XML content for empty elements.
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/settings/XMLFormattingOptions.java]: fixed

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/XMLFormatter.java [hunk 3]
```diff
@@ -280,131 +285,207 @@ private void format(DOMNode node) throws BadLocationException {
...
+		 * Format the given DOM element
+		 * 
+		 * @param element the DOM element to format.
+		 * 
+		 * @throws BadLocationException
+		 */
+		private void formatElement(DOMElement element) throws BadLocationException {
+			String tag = element.getTagName();
+			if (element.hasEndTag() && !element.hasStartTag()) {
+				// bad element which have not start tag (ex: <\root>)
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/XMLFormatter.java]: fixed

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/XMLFormatter.java [hunk 4]
```diff
@@ -551,10 +632,11 @@ public XMLFormatter(XMLExtensionsRegistry extensionsRegistry) {
...
 
 	/**
-	 * Returns a List containing a single TextEdit, containing the newly formatted changes
-	 * of the document.
+	 * Returns a List containing a single TextEdit, containing the newly formatted
+	 * changes of the document.
+	 * 
 	 * @param textDocument document to perform formatting on
-	 * @param range specified range in which formatting will be done
+	 * @param range        specified range in which formatting will be done
```
Comment (fbricon) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/XMLFormatter.java:639]: remove extra whitespace

**Comment by fbricon** [comment]
I find that if you have 
> "xml.format.preserveEmptyContent":true,
"xml.format.emptyElements":"collapse",

```xml
<foo>
    <bar>





        
    </bar>
</foo>
```

would ignore `"xml.format.preserveEmptyContent":true` and would yield

```xml
<foo>
    <bar />
</foo>
```

**Comment by angelozerr** [comment]
> I find that if you have

"xml.format.preserveEmptyContent":true,
"xml.format.emptyElements":"collapse",

fixed and tests are written for that

**Comment by fbricon** [comment]
actually we still have the same issue with "expand"

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/XMLFormatter.java (hunk 4 above)
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/XMLFormatter.java:670]: I cannot since it's a Eclipse format.

**Comment by angelozerr** [comment]
> actually we still have the same issue with "expand"

should be fixed.

**Comment by fbricon** [comment]
Test fails in https://ci.eclipse.org/lemminx/blue/organizations/jenkins/lemminx/detail/PR-644/1/tests
```

[ERROR] Failures: 

[ERROR]   XMLFormatterTest.collapseEmptyElements:2101->format:2207->format:2212->format:2246 expected: <<example att="hello">

  <X />

</example>> but was: <<example att="hello">

  <X />

</example>>

[INFO] 
```

**Pull request merged**